File: LeetCode/752. 打开转盘锁.rs
use std::collections::HashSet;
use std::collections::VecDeque;
impl Solution {
    pub fn open_lock(deadends: Vec<String>, target: String) -> i32 {
        let mut deadends: HashSet<String> = deadends.into_iter().collect();
        let mut q: VecDeque<String> = VecDeque::new();
        let mut visited: HashSet<String> = HashSet::new();
        q.push_back("0000".into());
        let mut step = 0;
        while q.len() != 0 {
            let l = q.len();
            for i in 0..l {
                let cur = q.pop_front().unwrap();
                if deadends.contains(&cur) {
                    continue;
                }
                if cur == target {
                    return step;
                }
                for next in 0..4 {
                    let minus = Solution::minus_one(&cur[..], next);
                    if !visited.contains(&minus)  {
                        q.push_back(minus.clone());
                        visited.insert(minus);
                    }
                    let plus = Solution::plus_one(&cur[..], next);
                    if !visited.contains(&plus) {
                        q.push_back(plus.clone());
                        visited.insert(plus);
                    }
                }
            }
            step += 1;
        }
        -1
    }
    fn minus_one(s: &str, idx: usize) -> String {
        s.chars().enumerate().map(|(i, c)| if i == idx {
            if c == '0' {
                '9'
            } else {
                ((c as u8) - 1) as char
            }
        } else {
            c
        }).collect()
    }
    fn plus_one(s: &str, idx: usize) -> String {
        s.chars().enumerate().map(|(i, c)| if i == idx {
            if c == '9' {
                '0'
            } else {
                ((c as u8) + 1) as char
            }
        } else {
            c
        }).collect()
    }
}

File: LeetCode/752. 打开转盘锁.ts
function openLock(deadends: string[], target: string): number {
    let q1: Set<string> = new Set();
    let q2: Set<string> = new Set();
    const deadendsSet = new Set(deadends);
    const visited: Set<string> = new Set();
    let step = 0;
    q1.add('0000');
    q2.add(target);
    const plusOne = (s: string, idx: number) => s.slice(0, idx) + (s[idx] === '9' ? '0' : String.fromCharCode(s.charCodeAt(idx) + 1)) + s.slice(idx + 1);
    const minusOne = (s: string, idx: number) => s.slice(0, idx) + (s[idx] === '0' ? '9' : String.fromCharCode(s.charCodeAt(idx) - 1)) + s.slice(idx + 1);
    while (q1.size !== 0 && q2.size !== 0) {
        const tmp: Set<string> = new Set();
        for(let x of q1) {
            if (deadendsSet.has(x)) {
                continue;
            }
            if (q2.has(x)) {
                return step;
            }
            visited.add(x);
            for (let i = 0; i < 4; i++) {
                const plus = plusOne(x, i);
                if (!visited.has(plus)) {
                    tmp.add(plus);
                }
                const minus = minusOne(x, i);
                if (!visited.has(minus)) {
                    tmp.add(minus);
                }
            }
        }
        step++;
        q1 = q2;
        q2 = tmp;
    }
    return -1;
};
